clients/src/main/java/org/apache/kafka/common/requests/PurgeRequest.java

```diff
@@ -0,0 +1,150 @@
+/**
+ * Licensed to the Apache Software Foundation (ASF) under one or more contributor license agreements. See the NOTICE
+ * file distributed with this work for additional information regarding copyright ownership. The ASF licenses this file
+ * to You under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with the
+ * License. You may obtain a copy of the License at
+ * <p>
+ * http://www.apache.org/licenses/LICENSE-2.0
+ * <p>
+ * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
+ * an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations under the License.
+ */
+
+package org.apache.kafka.common.requests;
+
+import org.apache.kafka.common.TopicPartition;
+import org.apache.kafka.common.protocol.ApiKeys;
+import org.apache.kafka.common.protocol.Errors;
+import org.apache.kafka.common.protocol.ProtoUtils;
+import org.apache.kafka.common.protocol.types.Struct;
+import org.apache.kafka.common.utils.CollectionUtils;
+import org.apache.kafka.common.utils.Utils;
+import java.nio.ByteBuffer;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+public class PurgeRequest extends AbstractRequest {
+
+    public static final long HIGH_WATERMARK = -1L;
+
+    // request level key names
+    private static final String TOPICS_KEY_NAME = "topics";
+    private static final String TIMEOUT_KEY_NAME = "timeout";
+
+    // topic level key names
+    private static final String TOPIC_KEY_NAME = "topic";
+    private static final String PARTITIONS_KEY_NAME = "partitions";
+
+    // partition level key names
+    private static final String PARTITION_KEY_NAME = "partition";
+    private static final String OFFSET_KEY_NAME = "offset";
+
+    private final int timeout;
+    private final Map<TopicPartition, Long> partitionOffsets;
+
+    public static class Builder extends AbstractRequest.Builder<PurgeRequest> {
+        private final int timeout;
+        private final Map<TopicPartition, Long> partitionOffsets;
+
+        public Builder(int timeout, Map<TopicPartition, Long> partitionOffsets) {
+            super(ApiKeys.PURGE);
+            this.timeout = timeout;
+            this.partitionOffsets = partitionOffsets;
+        }
+
+        @Override
+        public PurgeRequest build() {
+            short version = version();
+            return new PurgeRequest(timeout, partitionOffsets, version);
+        }
+
+        @Override
+        public String toString() {
+            StringBuilder builder = new StringBuilder();
```

Comment:
Is there any negative impact to use StringBuilder as compared to string concatenation? Using StringBuilder here allows us to have the same code style as `toString()` of other requests such as `ProduceRequest` and `LeaderAndIsrRequest`.